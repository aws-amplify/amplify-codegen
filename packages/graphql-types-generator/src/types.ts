export type Target = 'json' | 'swift' | 'ts' | 'typescript' | 'flow' | 'scala' | 'flow-modern' | 'angular';
