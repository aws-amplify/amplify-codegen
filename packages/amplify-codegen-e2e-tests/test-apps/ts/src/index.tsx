import React from 'react';
import ReactDOM from 'react-dom/client';
import { Amplify } from 'aws-amplify';
import App from './App';
import awsconfig from './aws-exports';

Amplify.configure(awsconfig);
const root = ReactDOM.createRoot(document.getElementById('root') as HTMLElement);
root.render(
  <React.StrictMode>
    <App />
  </React.StrictMode>,
);
