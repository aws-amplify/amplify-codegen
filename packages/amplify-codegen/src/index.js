const generateTypes = require('./commands/types');
const generateStatements = require('./commands/statements');
const generate = require('./commands/generateStatementsAndType');
const add = require('./commands/add');
const remove = require('./commands/remove');
const configure = require('./commands/configure');
const generateModels = require('./commands/models');
const { generateModelIntrospection, getModelIntrospection } = require('./commands/model-intropection');
const { isCodegenConfigured, switchToSDLSchema } = require('./utils');
const prePushAddGraphQLCodegenHook = require('./callbacks/prePushAddCallback');
const prePushUpdateGraphQLCodegenHook = require('./callbacks/prePushUpdateCallback');
const postPushGraphQLCodegenHook = require('./callbacks/postPushCallback');
const { executeAmplifyCommand, handleAmplifyEvent } = require('./amplify-plugin-index');
const { getCodegenConfig } = require('./codegen-config');

module.exports = {
  configure,
  generate,
  generateTypes,
  generateStatements,
  add,
  remove,
  prePushAddGraphQLCodegenHook,
  prePushUpdateGraphQLCodegenHook,
  postPushGraphQLCodegenHook,
  isCodegenConfigured,
  switchToSDLSchema,
  executeAmplifyCommand,
  handleAmplifyEvent,
  generateModels,
  generateModelIntrospection,
  getModelIntrospection,
  getCodegenConfig,
};

// temp change to trigger release
