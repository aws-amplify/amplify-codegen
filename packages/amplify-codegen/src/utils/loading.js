const fs = require('fs');
const { extname } = require('path');

function loadSchema(schemaPath) {
  if (!fs.existsSync(schemaPath)) {
    throw new Error(`Cannot find GraphQL schema file: ${schemaPath}`);
  }
  return fs.readFileSync(schemaPath, 'utf8');
}

function isSDLSchema(schemaPath) {
  if (extname(schemaPath) === '.json') {
    return false;
  }
  return true;
}

module.exports = { loadSchema, isSDLSchema };
