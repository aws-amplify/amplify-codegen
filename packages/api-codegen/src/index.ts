export type { GenerateModelsOptions } from './models';
export { generateModels } from './models';
