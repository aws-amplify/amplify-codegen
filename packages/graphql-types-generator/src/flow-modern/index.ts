export { generateSource, FlowGeneratedFile } from './codeGeneration';
