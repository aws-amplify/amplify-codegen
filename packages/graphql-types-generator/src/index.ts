import './polyfills';

export { default as generate, generateFromString } from './generate';
export type { Target } from './types';
export { getOutputFileName } from './utilities/getOutputFileName';
export { extractDocumentFromJavascript } from './loading';
