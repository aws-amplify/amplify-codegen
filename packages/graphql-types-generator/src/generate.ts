import { GraphQLSchema, DocumentNode, Source } from 'graphql';
import * as fs from 'fs-extra';
import * as path from 'path';
import * as rimraf from 'rimraf';

import { loadSchema, loadAndMergeQueryDocuments, parseSchema, parseAndMergeQueryDocuments } from './loading';
import { validateQueryDocument } from './validation';
import { compileToIR } from './compiler';
import { compileToLegacyIR } from './compiler/legacyIR';
import serializeToJSON from './serializeToJSON';
import { BasicGeneratedFileMap, isBasicGeneratedFileMap } from './utilities/CodeGenerator';
import { generateSource as generateSwiftSource } from './swift';
import { generateSource as generateTypescriptSource } from './typescript';
import { generateSource as generateFlowSource } from './flow';
import { generateSource as generateFlowModernSource } from './flow-modern';
import { generateSource as generateScalaSource } from './scala';
import { generateSource as generateAngularSource } from './angular';
import { hasS3Fields } from './utilities/complextypes';
import { Target } from './types';
import { getOutputFileName } from './utilities/getOutputFileName';

export default function generate(
  inputPaths: string[],
  schemaPath: string,
  outputPath: string,
  only: string,
  target: Target,
  tagName: string,
  options: any,
): void {
  generateFromFile(inputPaths, schemaPath, outputPath, only, target, tagName, options);
}

function generateFromFile(
  inputPaths: string[],
  schemaPath: string,
  outputPath: string,
  only: string,
  target: Target,
  tagName: string,
  options: any,
) {
  const schema = loadSchema(schemaPath);

  const document = loadAndMergeQueryDocuments(inputPaths, tagName);

  validateQueryDocument(schema, document);
  const multipleFiles = fs.existsSync(outputPath) && fs.statSync(outputPath).isDirectory();

  const output = generateForTarget(schema, document, only, target, multipleFiles, options);

  if (outputPath) {
    if (target === 'flow-modern') {
      // Group by output directory
      const filesByOutputDirectory: {
        [outputDirectory: string]: BasicGeneratedFileMap;
      } = {};

      // if target is flow-modern output must be BasicGeneratedFileMap
      Object.entries(output as BasicGeneratedFileMap).forEach(([filePath, file]) => {
        const outputDirectory = path.dirname(filePath);
        if (!filesByOutputDirectory[outputDirectory]) {
          filesByOutputDirectory[outputDirectory] = {
            [path.basename(filePath)]: file,
          };
        } else {
          filesByOutputDirectory[outputDirectory][path.basename(filePath)] = file;
        }
      });

      Object.keys(filesByOutputDirectory).forEach(outputDirectory => {
        writeGeneratedFiles(filesByOutputDirectory[outputDirectory], outputDirectory);
      });
    } else if (isBasicGeneratedFileMap(output)) {
      writeGeneratedFiles(output, outputPath);
    } else {
      fs.outputFileSync(outputPath, output);
    }
  } else {
    console.log(output);
  }
}

export function generateFromString(
  schema: string,
  introspection: boolean,
  queryDocuments: string,
  target: Target,
  multipleSwiftFiles: boolean,
  options: any,
): { [filepath: string]: string } {
  const graphqlSchema = parseSchema(schema, introspection);
  const document = parseAndMergeQueryDocuments([new Source(queryDocuments)]);
  validateQueryDocument(graphqlSchema, document);
  const output = generateForTarget(graphqlSchema, document, '', target, multipleSwiftFiles, options);

  if (isBasicGeneratedFileMap(output)) {
    return Object.entries(output)
      .map(([filepath, file]) => [filepath, file.output])
      .reduce((acc, [filepath, fileOutput]) => ({ ...acc, [filepath]: fileOutput }), {});
  }

  const filename = getOutputFileName('', target);
  return { [filename]: output };
}

export function generateForTarget(
  schema: GraphQLSchema,
  document: DocumentNode,
  only: string,
  target: Target,
  multipleFiles: boolean,
  options: any,
): string | BasicGeneratedFileMap {
  if (target === 'swift') {
    return generateTypesSwift(schema, document, only, multipleFiles, options);
  }
  if (target === 'flow-modern') {
    return generateTypesFlowModern(schema, document, options);
  }

  const context = compileToLegacyIR(schema, document, options);

  switch (target) {
    case 'json':
      return serializeToJSON(context);
    case 'ts':
    case 'typescript':
      return generateTypescriptSource(context);
    case 'flow':
      return generateFlowSource(context);
    case 'scala':
      return generateScalaSource(context, options);
    case 'angular':
      return generateAngularSource(context);
    default:
      throw new Error(`${target} is not supported.`);
  }
}

function generateTypesSwift(
  schema: GraphQLSchema,
  document: DocumentNode,
  only: string,
  multipleFiles: boolean,
  options: any,
): string | BasicGeneratedFileMap {
  options.addTypename = true;
  const context = compileToIR(schema, document, options);
  // Complex object suppport
  if (options.complexObjectSupport === 'auto') {
    options.addS3Wrapper = context.typesUsed.some(typesUsed => hasS3Fields(typesUsed));
  } else if (options.complexObjectSupport === 'yes') {
    options.addS3Wrapper = true;
  } else {
    options.addS3Wrapper = false;
  }

  const generator = generateSwiftSource(context, multipleFiles, only);

  if (multipleFiles) {
    return generator.generatedFiles;
  }
  return generator.output;
}

function generateTypesFlowModern(schema: GraphQLSchema, document: DocumentNode, options: any): BasicGeneratedFileMap {
  const context = compileToIR(schema, document, options);
  const generatedFiles = generateFlowModernSource(context);

  return generatedFiles;
}

function writeGeneratedFiles(generatedFiles: BasicGeneratedFileMap, outputDirectory: string) {
  // Clear all generated stuff to make sure there isn't anything
  // unnecessary lying around.
  rimraf.sync(outputDirectory);
  // Remake the output directory
  fs.mkdirSync(outputDirectory);

  for (const [fileName, generatedFile] of Object.entries(generatedFiles)) {
    fs.writeFileSync(path.join(outputDirectory, fileName), generatedFile.output);
  }
}
