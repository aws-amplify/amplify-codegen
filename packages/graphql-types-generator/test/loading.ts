import * as path from 'path';

import { loadAndMergeQueryDocuments } from '../src/loading';

describe('Validation', () => {
  test(`should extract gql snippet from javascript file`, () => {
    const inputPaths = [path.join(__dirname, '.', 'fixtures', 'starwars', 'gqlQueries.js')];

    const document = loadAndMergeQueryDocuments(inputPaths);
    // the document of type `DocumentNode` produced depends on the underlying OS and is hard to read.
    expect(document).toBeDefined;
  });
  test(`should throw when a file has invalid gql snippets`, () => {
    const inputPaths = [path.join(__dirname, '.', 'fixtures', 'misc', 'invalid-gqlQueries.js')];
    expect(() => {
      loadAndMergeQueryDocuments(inputPaths);
    }).toThrowError(
      `Could not parse graphql operations in ${path.join(
        'test',
        'fixtures',
        'misc',
        'invalid-gqlQueries.js',
      )}\n  Failed on : world and other words`,
    );
  });
});
