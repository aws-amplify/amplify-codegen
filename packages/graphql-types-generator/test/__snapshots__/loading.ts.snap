// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validation should extract gql snippet from javascript file 1`] = `
Object {
  "definitions": Array [
    Object {
      "directives": Array [],
      "kind": "OperationDefinition",
      "loc": Object {
        "end": 110,
        "start": 4,
      },
      "name": Object {
        "kind": "Name",
        "loc": Object {
          "end": 24,
          "start": 10,
        },
        "value": "HeroAndFriends",
      },
      "operation": "query",
      "selectionSet": Object {
        "kind": "SelectionSet",
        "loc": Object {
          "end": 110,
          "start": 44,
        },
        "selections": Array [
          Object {
            "alias": undefined,
            "arguments": Array [
              Object {
                "kind": "Argument",
                "loc": Object {
                  "end": 73,
                  "start": 56,
                },
                "name": Object {
                  "kind": "Name",
                  "loc": Object {
                    "end": 63,
                    "start": 56,
                  },
                  "value": "episode",
                },
                "value": Object {
                  "kind": "Variable",
                  "loc": Object {
                    "end": 73,
                    "start": 65,
                  },
                  "name": Object {
                    "kind": "Name",
                    "loc": Object {
                      "end": 73,
                      "start": 66,
                    },
                    "value": "episode",
                  },
                },
              },
            ],
            "directives": Array [],
            "kind": "Field",
            "loc": Object {
              "end": 105,
              "start": 51,
            },
            "name": Object {
              "kind": "Name",
              "loc": Object {
                "end": 55,
                "start": 51,
              },
              "value": "hero",
            },
            "selectionSet": Object {
              "kind": "SelectionSet",
              "loc": Object {
                "end": 105,
                "start": 75,
              },
              "selections": Array [
                Object {
                  "directives": Array [],
                  "kind": "FragmentSpread",
                  "loc": Object {
                    "end": 98,
                    "start": 84,
                  },
                  "name": Object {
                    "kind": "Name",
                    "loc": Object {
                      "end": 98,
                      "start": 87,
                    },
                    "value": "heroDetails",
                  },
                },
              ],
            },
          },
        ],
      },
      "variableDefinitions": Array [
        Object {
          "defaultValue": undefined,
          "directives": Array [],
          "kind": "VariableDefinition",
          "loc": Object {
            "end": 42,
            "start": 25,
          },
          "type": Object {
            "kind": "NamedType",
            "loc": Object {
              "end": 42,
              "start": 35,
            },
            "name": Object {
              "kind": "Name",
              "loc": Object {
                "end": 42,
                "start": 35,
              },
              "value": "Episode",
            },
          },
          "variable": Object {
            "kind": "Variable",
            "loc": Object {
              "end": 33,
              "start": 25,
            },
            "name": Object {
              "kind": "Name",
              "loc": Object {
                "end": 33,
                "start": 26,
              },
              "value": "episode",
            },
          },
        },
      ],
    },
    Object {
      "directives": Array [],
      "kind": "FragmentDefinition",
      "loc": Object {
        "end": 257,
        "start": 116,
      },
      "name": Object {
        "kind": "Name",
        "loc": Object {
          "end": 136,
          "start": 125,
        },
        "value": "heroDetails",
      },
      "selectionSet": Object {
        "kind": "SelectionSet",
        "loc": Object {
          "end": 257,
          "start": 150,
        },
        "selections": Array [
          Object {
            "alias": undefined,
            "arguments": Array [],
            "directives": Array [],
            "kind": "Field",
            "loc": Object {
              "end": 161,
              "start": 157,
            },
            "name": Object {
              "kind": "Name",
              "loc": Object {
                "end": 161,
                "start": 157,
              },
              "value": "name",
            },
            "selectionSet": undefined,
          },
          Object {
            "directives": Array [],
            "kind": "InlineFragment",
            "loc": Object {
              "end": 211,
              "start": 167,
            },
            "selectionSet": Object {
              "kind": "SelectionSet",
              "loc": Object {
                "end": 211,
                "start": 180,
              },
              "selections": Array [
                Object {
                  "alias": undefined,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "loc": Object {
                    "end": 204,
                    "start": 189,
                  },
                  "name": Object {
                    "kind": "Name",
                    "loc": Object {
                      "end": 204,
                      "start": 189,
                    },
                    "value": "primaryFunction",
                  },
                  "selectionSet": undefined,
                },
              ],
            },
            "typeCondition": Object {
              "kind": "NamedType",
              "loc": Object {
                "end": 179,
                "start": 174,
              },
              "name": Object {
                "kind": "Name",
                "loc": Object {
                  "end": 179,
                  "start": 174,
                },
                "value": "Droid",
              },
            },
          },
          Object {
            "directives": Array [],
            "kind": "InlineFragment",
            "loc": Object {
              "end": 252,
              "start": 217,
            },
            "selectionSet": Object {
              "kind": "SelectionSet",
              "loc": Object {
                "end": 252,
                "start": 230,
              },
              "selections": Array [
                Object {
                  "alias": undefined,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "loc": Object {
                    "end": 245,
                    "start": 239,
                  },
                  "name": Object {
                    "kind": "Name",
                    "loc": Object {
                      "end": 245,
                      "start": 239,
                    },
                    "value": "height",
                  },
                  "selectionSet": undefined,
                },
              ],
            },
            "typeCondition": Object {
              "kind": "NamedType",
              "loc": Object {
                "end": 229,
                "start": 224,
              },
              "name": Object {
                "kind": "Name",
                "loc": Object {
                  "end": 229,
                  "start": 224,
                },
                "value": "Human",
              },
            },
          },
        ],
      },
      "typeCondition": Object {
        "kind": "NamedType",
        "loc": Object {
          "end": 149,
          "start": 140,
        },
        "name": Object {
          "kind": "Name",
          "loc": Object {
            "end": 149,
            "start": 140,
          },
          "value": "Character",
        },
      },
    },
    Object {
      "directives": Array [],
      "kind": "OperationDefinition",
      "loc": Object {
        "end": 322,
        "start": 270,
      },
      "name": Object {
        "kind": "Name",
        "loc": Object {
          "end": 284,
          "start": 276,
        },
        "value": "HeroName",
      },
      "operation": "query",
      "selectionSet": Object {
        "kind": "SelectionSet",
        "loc": Object {
          "end": 322,
          "start": 285,
        },
        "selections": Array [
          Object {
            "alias": undefined,
            "arguments": Array [],
            "directives": Array [],
            "kind": "Field",
            "loc": Object {
              "end": 317,
              "start": 292,
            },
            "name": Object {
              "kind": "Name",
              "loc": Object {
                "end": 296,
                "start": 292,
              },
              "value": "hero",
            },
            "selectionSet": Object {
              "kind": "SelectionSet",
              "loc": Object {
                "end": 317,
                "start": 297,
              },
              "selections": Array [
                Object {
                  "alias": undefined,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "loc": Object {
                    "end": 310,
                    "start": 306,
                  },
                  "name": Object {
                    "kind": "Name",
                    "loc": Object {
                      "end": 310,
                      "start": 306,
                    },
                    "value": "name",
                  },
                  "selectionSet": undefined,
                },
              ],
            },
          },
        ],
      },
      "variableDefinitions": Array [],
    },
  ],
  "kind": "Document",
}
`;

exports[`Validation should throw a helpful message when a file has invalid gql snippets 1`] = `
"Could not parse graphql operations in test\\\\fixtures\\\\misc\\\\invalid-gqlQueries.js
  Failed on : world and other words"
`;
